api/external_auth.proto

```diff
@@ -0,0 +1,152 @@
+syntax = "proto3";
+
+// [#proto-status: draft]
+
+package envoy.api.v2;
+
+
+import "google/protobuf/struct.proto";
+import "google/protobuf/timestamp.proto";
+import "google/rpc/status.proto";
+
+// Authorization service.
+// Primarily responds with OK or NOT OK.
+service Authorization {
+   rpc Check(CheckRequest) returns (CheckResponse);
+}
+
+message CheckRequest {
+    // The name of the tenant for multi-tenant support.
+    string tenant_name = 1;
+
+    // The request attributes.
+    AttributeContext attributes = 2;
+}
+
+message CheckResponse {
+    // Status OK allows the request. Any other response indicates the request should be denied.
+    google.rpc.Status status = 1;
+}
+
+// An attribute is a piece of metadata that describes an activity on a network.
+// For example, the size of an HTTP request, or the status code of an HTTP response.
+//
+// Each attribute has a type and a name, which is logically defined as a proto message field.
+// The fields follow the Istio attribute vocabulary.
+// https://istio.io/docs/reference/config/mixer/attribute-vocabulary.html
+// The field type becomes the attribute type, and the field path becomes the attribute name.
+// For example, the attribute `source.ip` maps to field `AttributeContext.source.ip`.
+message AttributeContext {
+  // This message defines attributes for a node that handles a network request.
+  // The node can be either a service or an application that sends, forwards,
+  // or receives the request. Service peers should fill in the `service`,
+  // `principal`, and `labels` as appropriate.
+  message Peer {
+    // The IP address of the peer.
+    string ip = 1;
+
+    // The network port of the peer.
+    int64 port = 2;
+
+    // The canonical service name of the peer.
+    string service = 3;
+
+    // The labels associated with the peer.
+    // These could be pod labels for Kubernetes or tags for VMs.
+    // The source of the labels could be an X.509 certificate or other configuration.
+    map<string, string> labels = 4;
+
+    // The validated identity of this peer.
```

Comment:
The authenticated identity of this peer. If the server fully trusts the client, the server can apply policy directly against principal. Otherwise, the server can do additional validation based on labels and claims, such validation is known as "authentication policy".

We need to document the format well. For SPIFFE, the format must be ????. For Google accounts, it must be https://accounts.google.com/{userid}. etc etc.